homeassistant/components/media_player/mediaroom.py

```diff
@@ -9,134 +9,170 @@
-            ''.format(host),
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID)
-
-    add_devices(stbs)
+        await install_mediaroom_protocol(responses_callback=callback_notify)
+        _LOGGER.debug("Auto discovery installed")
 
 
 class MediaroomDevice(MediaPlayerDevice):
     """Representation of a Mediaroom set-up-box on the network."""
 
-    def __init__(self, name, host, optimistic=False, timeout=DEFAULT_TIMEOUT):
+    def set_state(self, mediaroom_state):
+        """Helper method to map pymediaroom states to HA states."""
+        from pymediaroom import State
+
+        state_map = {
+            State.OFF: STATE_OFF,
+            State.STANDBY: STATE_STANDBY,
+            State.PLAYING_LIVE_TV: STATE_PLAYING,
+            State.PLAYING_RECORDED_TV: STATE_PLAYING,
+            State.PLAYING_TIMESHIFT_TV: STATE_PLAYING,
+            State.STOPPED: STATE_PAUSED,
+            State.UNKNOWN: STATE_UNAVAILABLE
+            }
+
+        self._state = state_map[mediaroom_state]
+
+    def __init__(self, host, device_id, optimistic=False,
+                 timeout=DEFAULT_TIMEOUT):
         """Initialize the device."""
         from pymediaroom import Remote
 
-        self.stb = Remote(host, timeout=timeout)
-        _LOGGER.info(
-            "Found %s at %s%s", name, host,
-            " - I'm optimistic" if optimistic else "")
-        self._name = name
-        self._is_standby = not optimistic
-        self._current = None
+        self.host = host
+        self.stb = Remote(host)
+        _LOGGER.info("Found STB at %s%s", host,
+                     " - I'm optimistic" if optimistic else "")
+        self._channel = None
         self._optimistic = optimistic
-        self._state = STATE_STANDBY
+        self._state = STATE_PLAYING if optimistic else STATE_STANDBY
+        self._name = '{} {}'.format(PLATFORM, device_id)
+        if device_id:
+            self._unique_id = device_id
+        else:
+            self._unique_id = self.stb.device_id
 
-    def update(self):
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    async def async_added_to_hass(self):
         """Retrieve latest state."""
-        if not self._optimistic:
-            self._is_standby = self.stb.get_standby()
-        if self._is_standby:
-            self._state = STATE_STANDBY
-        elif self._state not in [STATE_PLAYING, STATE_PAUSED]:
-            self._state = STATE_PLAYING
-        _LOGGER.debug(
-            "%s(%s) is [%s]",
-            self._name, self.stb.stb_ip, self._state)
-
-    def play_media(self, media_type, media_id, **kwargs):
+        def notify_received(notify):
+            """Process STB state from NOTIFY message."""
+            self.set_state(self.stb.notify_callback(notify))
+            _LOGGER.debug("STB(%s) is [%s]", self.host, self._state)
+            self.async_schedule_update_ha_state()
```

Comment:
You cannot call async methods from a sync context

Reply:
Isn't notify_received called from an async context (being a callback installed by async_dispatcher_connect) ?

The sync methods are just translators...

Reply:
This method is not async and thus will not be called from an async context. If you want this method to be async, either add the callback decorator or make it an async function.

https://home-assistant.io/developers/asyncio_categorizing_functions/